src/solvers/petsc_nonlinear_solver.C

```diff
@@ -943,6 +943,12 @@ PetscNonlinearSolver<T>::solve (SparseMatrix<T> &  pre_in,  // System Preconditi
   ierr = SNESGetLinearSolveIterations(_snes, &_n_linear_iterations);
   LIBMESH_CHKERR(ierr);
 
+  // Enforce constraints exactly now that the solve is done.  We have
+  // been enforcing them on the current_local_solution during the
+  // solve, but now need to be sure they are enforced on the parallel
+  // solution vector as well.
+  this->system().get_dof_map().enforce_constraints_exactly(this->system());
+
```

Comment:
Why do we need to enforce it after solve? We manually to change the solution to match prescribed constraints?

Reply:
OK, I believe I see the whole picture now.  I will add more changes, and then we do not need this anymore.

It should be very easy